File: frontend/src/services/authService.js
import { request } from './api';

const SESSION_STORAGE_KEY = 'agro_smart_session';
const USERS_REGISTRY_KEY = 'agro_smart_users';

export const normalizeRole = (roleStr) => {
  const clean = String(roleStr || '').toLowerCase().trim().replace(/[- ]/g, '_');
  if (clean.includes('admin')) return 'admin';
  if (clean.includes('owner') || clean.includes('machine')) return 'machine_owner';
  return 'farmer';
};

export const getRoleDisplayName = (roleStr) => {
  const norm = normalizeRole(roleStr);
  if (norm === 'admin') return 'Admin';
  if (norm === 'machine_owner') return 'Machinery Owner';
  return 'Farmer';
};

// Pre-seeded demo credentials for instant hackathon testing
export const DEMO_CREDENTIALS = {
  farmer: {
    identifier: '9876543210',
    email: 'farmer@agro-smart.com',
    password: 'Farmer@123',
    role: 'farmer',
    user_type: 'Farmer',
    name: 'Rameshwar Patel',
    state: 'Bihar',
    district: 'Patna'
  },
  owner: {
    identifier: '9876543211',
    email: 'owner@agro-smart.com',
    password: 'Owner@123',
    role: 'machine_owner',
    user_type: 'Machinery Owner',
    name: 'Suresh Singh Machinery',
    state: 'Maharashtra',
    district: 'Pune'
  },
  admin: {
    identifier: '9876500000',
    email: 'admin@agro-smart.com',
    password: 'Admin@123',
    role: 'admin',
    user_type: 'Admin',
    name: 'AGRO-SMART System Admin',
    state: 'Maharashtra',
    district: 'Pune'
  }
};

// Simple hash for demo client-side fallback validation (never store plaintext)
const simpleHash = (str) => {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash |= 0;
  }
  return 'h_' + Math.abs(hash).toString(16);
};

export const authService = {
  /**
   * Primary Login: POST /api/auth/login with local demo fallback.
   */
  async login(identifier, password, rememberMe = true) {
    const cleanId = String(identifier || '').trim().toLowerCase();
    if (!cleanId) throw new Error('Please enter your mobile number or email.');
    if (!password) throw new Error('Please enter your password.');

    // 1. Try Backend API
    try {
      const data = await request('/auth/login', {
        method: 'POST',
        body: JSON.stringify({ identifier: cleanId, password })
      });

      if (data && data.user) {
        const normUser = this._normalizeUser(data.user);
        this.saveSession(normUser, data.token, rememberMe);
        return { user: normUser, token: data.token, is_demo_session: true };
      }
    } catch (err) {
      console.warn('[AuthService] Backend login error (using local fallback):', err.message);
      if (err.message && err.message.includes('account has been disabled')) {
        throw err;
      }
    }

    // 2. Client-side Demo Fallback
    let matched = null;
    if (cleanId === DEMO_CREDENTIALS.farmer.identifier || cleanId === DEMO_CREDENTIALS.farmer.email) {
      if (password === DEMO_CREDENTIALS.farmer.password) {
        matched = {
          id: 'usr-demo-farmer-01',
          name: DEMO_CREDENTIALS.farmer.name,
          email: DEMO_CREDENTIALS.farmer.email,
          phone: DEMO_CREDENTIALS.farmer.identifier,
          role: 'farmer',
          user_type: 'Farmer',
          state: DEMO_CREDENTIALS.farmer.state,
          district: DEMO_CREDENTIALS.farmer.district,
          avatar: '👨‍🌾',
          status: 'Active'
        };
      } else {
        throw new Error('Invalid email/mobile or password.');
      }
    } else if (cleanId === DEMO_CREDENTIALS.owner.identifier || cleanId === DEMO_CREDENTIALS.owner.email) {
      if (password === DEMO_CREDENTIALS.owner.password) {
        matched = {
          id: 'usr-demo-owner-02',
          name: DEMO_CREDENTIALS.owner.name,
          email: DEMO_CREDENTIALS.owner.email,
          phone: DEMO_CREDENTIALS.owner.identifier,
          role: 'machine_owner',
          user_type: 'Machinery Owner',
          state: DEMO_CREDENTIALS.owner.state,
          district: DEMO_CREDENTIALS.owner.district,
          avatar: '🚜',
          status: 'Active'
        };
      } else {
        throw new Error('Invalid email/mobile or password.');
      }
    } else if (cleanId === DEMO_CREDENTIALS.admin.identifier || cleanId === DEMO_CREDENTIALS.admin.email) {
      if (password === DEMO_CREDENTIALS.admin.password) {
        matched = {
          id: 'usr-demo-admin-00',
          name: DEMO_CREDENTIALS.admin.name,
          email: DEMO_CREDENTIALS.admin.email,
          phone: DEMO_CREDENTIALS.admin.identifier,
          role: 'admin',
          user_type: 'Admin',
          state: DEMO_CREDENTIALS.admin.state,
          district: DEMO_CREDENTIALS.admin.district,
          avatar: '🛡️',
          status: 'Active'
        };
      } else {
        throw new Error('Invalid email/mobile or password.');
      }
    } else {
      // Check registered users in local storage
      const registered = this.getRegisteredDemoUsers();
      const found = registered.find(u => u.email.toLowerCase() === cleanId || u.phone === cleanId);
      if (found) {
        if (found.status === 'Disabled') {
          throw new Error('This account has been disabled by the administrator.');
        }
        if (found.passwordHash === simpleHash(password)) {
          matched = this._normalizeUser(found);
        } else {
          throw new Error('Invalid email/mobile or password.');
        }
      }
    }

    if (matched) {
      const token = `demo-tok-${Date.now()}`;
      this.saveSession(matched, token, rememberMe);
      return { user: matched, token, is_demo_session: true };
    }

    throw new Error('Account not found.');
  },

  /**
   * Primary Register: POST /api/auth/register with local demo fallback.
   */
  async register(formData) {
    const rawRole = formData.role || formData.user_type || 'farmer';
    let normRole = normalizeRole(rawRole);
    if (normRole === 'admin') normRole = 'farmer'; // No public admin registration

    const cleanData = {
      name: (formData.name || formData.fullName || '').trim(),
      phone: (formData.phone || formData.mobileNumber || '').trim(),
      email: (formData.email || '').trim().toLowerCase(),
      password: formData.password || '',
      role: normRole,
      user_type: getRoleDisplayName(normRole),
      state: formData.state || 'Maharashtra',
      district: formData.district || 'Pune'
    };

    // 1. Try Backend API
    try {
      const res = await request('/auth/register', {
        method: 'POST',
        body: JSON.stringify(cleanData)
      });
      if (res && res.user) {
        const normUser = this._normalizeUser(res.user);
        this.saveSession(normUser, res.token, true);
        return { user: normUser, token: res.token, is_demo_session: true };
      }
    } catch (err) {
      console.warn('[AuthService] Backend register error (using local fallback):', err.message);
      if (err.message && err.message.includes('already exists')) {
        throw err;
      }
    }

    // 2. Client-side Demo Fallback
    const registered = this.getRegisteredDemoUsers();
    if (registered.some(u => u.email === cleanData.email)) {
      throw new Error('An account with this email address already exists. Please log in.');
    }
    if (registered.some(u => u.phone === cleanData.phone)) {
      throw new Error('An account with this mobile number already exists. Please log in.');
    }

    const newUser = {
      id: `usr-${Date.now()}`,
      name: cleanData.name,
      email: cleanData.email,
      phone: cleanData.phone,
      passwordHash: simpleHash(cleanData.password),
      role: cleanData.role,
      user_type: cleanData.user_type,
      state: cleanData.state,
      district: cleanData.district,
      avatar: cleanData.role === 'machine_owner' ? '🚜' : '👨‍🌾',
      status: 'Active',
      created_at: new Date().toISOString().replace('T', ' ').substring(0, 19)
    };

    registered.push(newUser);
    this.saveRegisteredDemoUsers(registered);

    const safeUser = this._normalizeUser(newUser);
    const token = `demo-tok-${Date.now()}`;
    this.saveSession(safeUser, token, true);
    return { user: safeUser, token, is_demo_session: true };
  },

  /**
   * Save session safely to localStorage or sessionStorage.
   * NEVER stores raw password.
   */
  saveSession(user, token, rememberMe = true) {
    if (!user) return;
    const safeSession = {
      user: this._normalizeUser(user),
      token: token || `demo-tok-${Date.now()}`,
      savedAt: Date.now(),
      is_demo_session: true
    };
    const jsonStr = JSON.stringify(safeSession);
    try {
      if (rememberMe) {
        localStorage.setItem(SESSION_STORAGE_KEY, jsonStr);
        sessionStorage.removeItem(SESSION_STORAGE_KEY);
      } else {
        sessionStorage.setItem(SESSION_STORAGE_KEY, jsonStr);
        localStorage.removeItem(SESSION_STORAGE_KEY);
      }
    } catch (err) {
      console.warn('[AuthService] Storage save error:', err);
    }
  },

  /**
   * Restore safe session with fallback recovery.
   */
  getSession() {
    try {
      const raw = localStorage.getItem(SESSION_STORAGE_KEY) || sessionStorage.getItem(SESSION_STORAGE_KEY);
      if (!raw) return null;
      const parsed = JSON.parse(raw);
      if (parsed && parsed.user) {
        parsed.user = this._normalizeUser(parsed.user);
        return parsed;
      }
      return null;
    } catch (err) {
      console.warn('[AuthService] Session parse error:', err);
      return null;
    }
  },

  /**
   * Clears session storage on logout.
   */
  async logout() {
    try {
      const session = this.getSession();
      if (session && session.token) {
        await request('/auth/logout', {
          method: 'POST',
          headers: { Authorization: `Bearer ${session.token}` }
        }).catch(() => {});
      }
    } finally {
      localStorage.removeItem(SESSION_STORAGE_KEY);
      sessionStorage.removeItem(SESSION_STORAGE_KEY);
    }
  },

  getRegisteredDemoUsers() {
    try {
      const raw = localStorage.getItem(USERS_REGISTRY_KEY);
      return raw ? JSON.parse(raw) : [];
    } catch {
      return [];
    }
  },

  saveRegisteredDemoUsers(users) {
    try {
      localStorage.setItem(USERS_REGISTRY_KEY, JSON.stringify(users));
    } catch (err) {
      console.warn('[AuthService] User registry save error:', err);
    }
  },

  _normalizeUser(user) {
    if (!user) return null;
    const normRole = normalizeRole(user.role || user.user_type);
    return {
      id: user.id || `usr-${Date.now()}`,
      name: user.name || 'User',
      email: user.email || '',
      phone: user.phone || '',
      role: normRole,
      user_type: getRoleDisplayName(normRole),
      state: user.state || 'Maharashtra',
      district: user.district || 'Pune',
      avatar: user.avatar || (normRole === 'admin' ? '🛡️' : (normRole === 'machine_owner' ? '🚜' : '👨‍🌾')),
      status: user.status || 'Active',
      created_at: user.created_at || new Date().toISOString().replace('T', ' ').substring(0, 19)
    };
  }
};

File: backend/services/auth_service.py
"""
AGRO-SMART Authentication Service
Handles user registration, credential verification, password hashing, and session tokens.
Supports local demo mode with in-memory stores and Supabase database fallback.
"""
import hashlib
import time
import uuid
import os
from models import supabase_client

# Salt for demo password hashing
AUTH_SALT = "agro_smart_secret_salt_2026"

def _hash_password(password: str) -> str:
    """Generates salted SHA-256 hash for secure credential storage."""
    return hashlib.sha256(f"{password}:{AUTH_SALT}".encode()).hexdigest()

# Pre-seeded Hackathon Demo Accounts
INITIAL_DEMO_USERS = [
    {
        "id": "usr-demo-farmer-01",
        "name": "Rameshwar Patel",
        "email": "farmer@agro-smart.com",
        "phone": "9876543210",
        "password_hash": _hash_password("Farmer@123"),
        "user_type": "Farmer",
        "state": "Bihar",
        "district": "Patna",
        "avatar": "👨‍🌾",
        "created_at": "2026-08-01 10:00:00"
    },
    {
        "id": "usr-demo-owner-02",
        "name": "Suresh Singh Machinery",
        "email": "owner@agro-smart.com",
        "phone": "9876543211",
        "password_hash": _hash_password("Owner@123"),
        "user_type": "Machinery Owner",
        "state": "Maharashtra",
        "district": "Pune",
        "avatar": "🚜",
        "created_at": "2026-08-01 11:30:00"
    }
]

# In-memory user database for demo session resilience
USERS_STORE = list(INITIAL_DEMO_USERS)
SESSIONS_STORE = {}

def register_user(data: dict):
    """
    Registers a new user (Farmer or Machinery Owner).
    Validates required fields, checks for duplicate email/phone, hashes password.
    """
    name = (data.get("name") or data.get("fullName") or "").strip()
    phone = (data.get("phone") or data.get("mobileNumber") or "").strip()
    email = (data.get("email") or "").strip().lower()
    password = data.get("password") or ""
    user_type = data.get("user_type") or data.get("userType") or "Farmer"
    state = (data.get("state") or "Maharashtra").strip()
    district = (data.get("district") or "Pune").strip()

    if not name:
        return None, "Full name is required."
    if not phone or len(phone) < 10:
        return None, "Valid 10-digit mobile number is required."
    if not email or "@" not in email:
        return None, "Valid email address is required."
    if not password or len(password) < 6:
        return None, "Password must be at least 6 characters long."
    if user_type not in ["Farmer", "Machinery Owner"]:
        user_type = "Farmer"

    # Check for duplicate phone or email in local memory
    for u in USERS_STORE:
        if u["phone"] == phone:
            return None, "An account with this mobile number already exists. Please log in."
        if u["email"] == email:
            return None, "An account with this email address already exists. Please log in."

    new_user = {
        "id": f"usr-{uuid.uuid4().hex[:10]}",
        "name": name,
        "email": email,
        "phone": phone,
        "password_hash": _hash_password(password),
        "user_type": user_type,
        "state": state,
        "district": district,
        "avatar": "🚜" if user_type == "Machinery Owner" else "👨‍🌾",
        "created_at": time.strftime("%Y-%m-%d %H:%M:%S")
    }

    # Save to Supabase if live database is connected
    try:
        if supabase_client:
            record = {
                "id": new_user["id"],
                "name": name,
                "email": email,
                "phone": phone,
                "user_type": user_type,
                "state": state,
                "district": district
            }
            supabase_client.table("users").insert(record).execute()
    except Exception as e:
        print(f"[Supabase] User insert notice (using in-memory store): {e}")

    USERS_STORE.append(new_user)

    # Generate session token and public profile
    token = f"demo-tok-{uuid.uuid4().hex}"
    SESSIONS_STORE[token] = new_user["id"]

    public_user = _sanitize_user(new_user)
    return {
        "user": public_user,
        "token": token,
        "is_demo_session": True,
        "message": "Account created successfully"
    }, None

def authenticate_user(identifier: str, password: str):
    """
    Authenticates user via Mobile Number or Email + Password.
    Returns sanitized user profile and session token.
    """
    clean_id = (identifier or "").strip().lower()
    if not clean_id:
        return None, "Please enter your mobile number or email."
    if not password:
        return None, "Please enter your password."

    hashed_input = _hash_password(password)

    # Search user store
    matched_user = None
    for u in USERS_STORE:
        if u["email"].lower() == clean_id or u["phone"] == clean_id:
            matched_user = u
            break

    if not matched_user:
        return None, "No account found with this mobile number or email. Please register."

    if matched_user["password_hash"] != hashed_input:
        return None, "Invalid password. Please check and try again."

    # Generate session
    token = f"demo-tok-{uuid.uuid4().hex}"
    SESSIONS_STORE[token] = matched_user["id"]

    public_user = _sanitize_user(matched_user)
    return {
        "user": public_user,
        "token": token,
        "is_demo_session": True,
        "message": f"Welcome back, {matched_user['name']}!"
    }, None

def get_user_by_token(token: str):
    """Retrieves authenticated user profile from active session token."""
    if not token or token not in SESSIONS_STORE:
        return None, "Session expired or invalid. Please log in again."

    user_id = SESSIONS_STORE[token]
    for u in USERS_STORE:
        if u["id"] == user_id:
            return _sanitize_user(u), None

    return None, "User not found."

def logout_token(token: str):
    """Invalidates active session token."""
    if token and token in SESSIONS_STORE:
        del SESSIONS_STORE[token]
    return True

def _sanitize_user(user: dict) -> dict:
    """Removes sensitive password hashes before returning user profile to frontend."""
    return {
        "id": user["id"],
        "name": user["name"],
        "email": user["email"],
        "phone": user["phone"],
        "user_type": user["user_type"],
        "state": user["state"],
        "district": user["district"],
        "avatar": user.get("avatar", "👨‍🌾"),
        "created_at": user.get("created_at")
    }
